// odies.database/src/YAMLTable.js
const EventEmitter = require("events").EventEmitter;
let db;

class YAML extends EventEmitter {

    constructor(file, options) {
        super();
        if (!file) return console.error('[YAML] : Bir dosya ismi girmelisiniz.')
        if (file.endsWith('.yaml')) return console.error('[YAML] : Sadece dosya ismini giriniz.')
        if (!options["table"]) return console.error('[YAML] : Bir "table" ismi giriniz.')

        this.emit("ready");

        db = require('v2db')({
			name: `${file}_${options["table"]}`, // varsayılan: db
			seperator: '.', // varsayılan: .
			language: 'tr', // varsayılan: en
			yaml: true, // varsayılan: false
});
    }

    async set(data, value) {
        if (!data) return console.error('[YAML] : Bir "data" giriniz.')
        if (!value) return console.error('[YAML] : Bir "value" giriniz.')

        await db.set(data, value)
        return await db.get(data)
    }

    async delete(data) {
        if (!data) return console.error('[YAML] : Bir "data" giriniz.')

        return await db.delete(data)
    }

    async deleteAll() {
        return await db.deleteAll()
    }

    async fetch(data) {
        if (!data) return console.error('[YAML] : Bir "data" giriniz.')

        return await db.get(data)
    }

    async get(data) {
        if (!data) return console.error('[YAML] : Bir "data" giriniz.')

        return await db.get(data)
    }

    async add(data, number) {
        if (!data) return console.error('[YAML] : Bir "data" giriniz.')
        if (!number) return console.error('[YAML] : Bir "number" giriniz.')

        await db.add(data, number)
        return await db.get(data)
    }

    async subtract(data, number) {
        if (!data) return console.error('[YAML] : Bir "data" giriniz.')
        if (!number) return console.error('[YAML] : Bir "number" giriniz.')

        await db.subtract(data, number)
        return await db.get(data)
    }

    async push(data, value) {
        if (!data) return console.error('[YAML] : Bir "data" giriniz.');
        if (!value) return console.error('[YAML] : Bir "value" giriniz.');

        await db.push(data, value);
        return await db.get(data);
    }

    async unpush(data, value) {
        if (!data) return console.error('[YAML] : Bir "data" giriniz.');
        if (!value) return console.error('[YAML] : Bir "value" giriniz.');

        await db.unpush(data, value);
        return await db.get(data);
    }

    async all() {
        return await db.all()
    }

    async move(database, data) {
        if (database["database"] !== 'quick.db') {
            return console.error('[YAML] : "quick.db" modülünü giriniz.')
        }

        const All = database["database"].all()
        All.map(x => {
            db.set(x.data, x.ID)
        })

        return true;
    }

}

module.exports = YAML;
